dcap-ql/src/quote.rs

```diff
@@ -332,6 +452,84 @@ impl<'a> Qe3CertDataPpid<'a> {
     }
 }
 
+pub trait Quote3SignatureEcdsaP256Verifier {
+    /// Verify the platform certification data.
+    ///
+    /// The certification data is in `quote3signature.certification_data()`.
+    ///
+    /// On success, should return the platform certification public key (PCK).
+    // TODO: pass a container for the certification type and data instead of
+    // the whole signature structure.
+    fn verify_certification_data<'a>(&mut self, quote3signature: &'a Quote3SignatureEcdsaP256) -> Result<Cow<'a, [u8]>>;
+
+    /// Verify the quoting enclave (QE3).
+    fn verify_qe3(&mut self, qe3_report: &[u8], authentication_data: &[u8]) -> Result<()>;
+}
+
+pub trait Quote3SignatureVerify<'a>: Quote3Signature<'a> {
+    type TrustRoot;
+
+    fn verify(&self, quote: &[u8], root_of_trust: Self::TrustRoot) -> Result<()>;
+}
+
+#[cfg(feature = "verify")]
+impl<'a> Quote3SignatureVerify<'a> for Quote3SignatureEcdsaP256<'a> {
+    type TrustRoot = &'a mut dyn Quote3SignatureEcdsaP256Verifier;
+
+    fn verify(&self, quote: &[u8], verifier: Self::TrustRoot) -> Result<()> {
+        // Verify certification data
+        let pck = verifier.verify_certification_data(self)?;
+
+        // verify qe3_signature against qe3_report/authentication_data using pck
+        //   verify QE report signature signed by pck
+        let sig = self.qe3_signature();
+        let sig = get_ecdsa_sig_der(sig)?;
+        let data = self.qe3_report();
+        let mut hash = [0u8; 32];
+        Md::hash(hash::Type::Sha256, &data, &mut hash)?;
+        let pck = CString::new(pck)?;
+        let mut pck = Certificate::from_pem(pck.as_bytes_with_nul())?;
+        let pck_pk = pck.public_key_mut();
```

Comment:
I guess I wasn't that explicit, but I intended you to do `Pk::from_public_key` here. And that `verify_certification_data` returns a key in DER format.